File: kattis/Exercise6/islandhop.py
import math

def find(parent, x):
    if parent[x] != x:
        parent[x] = find(parent, parent[x])
    return parent[x]

def union(parent, rank, a, b):
    rootA = find(parent, a)
    rootB = find(parent, b)
    if rootA != rootB:
        if rank[rootA] < rank[rootB]:
            parent[rootA] = rootB
        elif rank[rootA] > rank[rootB]:
            parent[rootB] = rootA
        else:
            parent[rootB] = rootA
            rank[rootA] += 1
        return True
    return False

def solve():
    t = int(input().strip())  # number of test cases
    for _ in range(t):
        m = int(input().strip())  # number of islands
        coords = [tuple(map(float, input().split())) for _ in range(m)]

        edges = []
        for i in range(m):
            x1, y1 = coords[i]
            for j in range(i+1, m):
                x2, y2 = coords[j]
                dist = math.hypot(x2 - x1, y2 - y1)
                edges.append((dist, i, j))

        edges.sort(key=lambda e: e[0])

        parent = list(range(m))
        rank = [0]*m
        mst_length = 0.0
        edges_used = 0

        for dist, a, b in edges:
            if union(parent, rank, a, b):
                mst_length += dist
                edges_used += 1
                if edges_used == m - 1:
                    break

        print(f"{mst_length:.6f}")

File: kattis/Exercise6/islandhop.cpp
#include <bits/stdc++.h>
using namespace std;

struct UnionFind {
    vector<int> parent, rank_;
    UnionFind(int n) : parent(n), rank_(n, 0) {
        for(int i = 0; i < n; i++) parent[i] = i;
    }
    
    int findSet(int v) {
        if(parent[v] != v) parent[v] = findSet(parent[v]);
        return parent[v];
    }
    
    bool unionSet(int a, int b) {
        a = findSet(a), b = findSet(b);
        if(a == b) return false;
        if(rank_[a] < rank_[b]) swap(a, b);
        parent[b] = a;
        if(rank_[a] == rank_[b]) rank_[a]++;
        return true;
    }
};

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;  // number of test cases
    cin >> T;
    while(T--){
        int m; // number of islands
        cin >> m;
        
        vector<pair<double,double>> coords(m);
        for(int i = 0; i < m; i++){
            cin >> coords[i].first >> coords[i].second;
        }
        
        vector<tuple<double,int,int>> edges;
        edges.reserve((long long)m*(m-1)/2);  // safe reservation for up to m*(m-1)/2 edges
        
        for(int i = 0; i < m; i++){
            for(int j = i+1; j < m; j++){
                double dx = coords[i].first - coords[j].first;
                double dy = coords[i].second - coords[j].second;
                double dist = sqrt(dx*dx + dy*dy);
                edges.push_back({dist, i, j});
            }
        }
        
        sort(edges.begin(), edges.end(),
             [](auto &a, auto &b){
                 return get<0>(a) < get<0>(b);
             });
        
        UnionFind uf(m);
        double mst_length = 0.0;
        int edges_used = 0;
        
        for(auto &e : edges){
            double dist;
            int a, b;
            tie(dist, a, b) = e;
            if(uf.unionSet(a,b)){
                mst_length += dist;
                edges_used++;
                if(edges_used == m-1) break;
            }
        }
        
        cout << fixed << setprecision(6) << mst_length << "\n";
    }
    return 0;
}
